src/Common/src/TypeSystem/Interop/IL/Marshaller.cs

```diff
@@ -856,44 +1186,473 @@ protected override void EmitMarshalArgumentManagedToNative()
+            }
+            _elementMarshaller.In = (direction == MarshalDirection);
+            _elementMarshaller.Out = !In;
+            _elementMarshaller.MarshalDirection = MarshalDirection;
+
+            return _elementMarshaller;
+        }
+
+        protected virtual void EmitElementCount(ILCodeStream codeStream, MarshalDirection direction)
+        {
+            if (direction == MarshalDirection.Forward)
+            {
+                // In forward direction we skip whatever is passed through SizeParamIndex, becaus the
+                // size of the managed array is already known
+                codeStream.EmitLdLoc(_vManaged);
+                codeStream.Emit(ILOpcode.ldlen);
+                codeStream.Emit(ILOpcode.conv_i4);
+
+            }
+            else if (MarshalDirection == MarshalDirection.Forward
+                    && MarshallerType == MarshallerType.Argument
+                    && !Return
+                    && !IsByRef)
+            {
+                EmitElementCount(codeStream, MarshalDirection.Forward);
+            }
+            else
+            { 
+
+                uint? sizeParamIndex = PInvokeParameterMetadata.MarshalAsDescriptor.SizeParamIndex;
+                uint? sizeConst = PInvokeParameterMetadata.MarshalAsDescriptor.SizeConst;
+
+                if (sizeConst.HasValue)
+                {
+                    codeStream.EmitLdc((int)sizeConst.Value);
+                }
+
+                if (sizeParamIndex.HasValue)
+                {
+                    uint index = sizeParamIndex.Value;
+
+                    if (index < 0 || index >= Marshallers.Length -1)
+                    {
+                        throw new InvalidProgramException("Invalid SizeParamIndex, must be between 0 and parameter count");
+                    }
+
+                    //zero-th index is for return type
+                    index++;
+                    var indexType = Marshallers[index].ManagedParameterType;
+                    bool isByRef = indexType.IsByRef;
+
+                    if (isByRef)
+                    {
+                        indexType = ((ByRefType)indexType).ParameterType;
+                    }
+
+                    switch (indexType.Category)
+                    {
+                        case TypeFlags.Byte:
+                        case TypeFlags.SByte:
+                        case TypeFlags.Int16:
+                        case TypeFlags.UInt16:
+                        case TypeFlags.Int32:
+                        case TypeFlags.UInt32:
+                        case TypeFlags.Int64:
+                        case TypeFlags.UInt64:
+                            break;
+                        default:
+                            throw new InvalidProgramException("Invalid SizeParamIndex, parameter must be  of type int/uint");
+                    }
+
+                    codeStream.EmitLdArg(Marshallers[index].PInvokeParameterMetadata.Index - 1);
```

Comment:
I know passing the marshaller array to each marshallers is ugly. However, this ParameterMetadata is used from other places(eg. default value on enum) and adding the SizeParamIndex parameter there and validating it may not be the expected. I will leave it like this for now.